blockchain/peer.go

```diff
@@ -0,0 +1,185 @@
+		height:     height,
+		blocks:     make(map[int64]*types.Block, maxRequestsPerPeer),
+		logger:     log.NewNopLogger(),
+		errFunc:    errFunc,
+		parameters: parameters,
+	}
+}
+
+func bpPeerDefaultParams() *bpPeerParams {
+	return &bpPeerParams{
+		// Timeout for a peer to respond to a block request.
+		peerTimeout: 15 * time.Second,
+
+		// Minimum recv rate to ensure we're receiving blocks from a peer fast
+		// enough. If a peer is not sending data at at least that rate, we
+		// consider them to have timedout and we disconnect.
+		//
+		// Assuming a DSL connection (not a good choice) 128 Kbps (upload) ~ 15 KB/s,
+		// sending data across atlantic ~ 7.5 KB/s.
+		minRecvRate: int64(7680),
+
+		// Monitor parameters
+		peerSampleRate: time.Second,
+		peerWindowSize: 40 * time.Second,
+	}
+}
+
+func (peer *bpPeer) ID() p2p.ID {
+	return peer.id
+}
+
+func (peer *bpPeer) SetLogger(l log.Logger) {
+	peer.logger = l
+}
+
+func (peer *bpPeer) GetHeight() int64 {
+	return peer.height
+}
+
+func (peer *bpPeer) SetHeight(height int64) {
+	peer.height = height
+}
+
+func (peer *bpPeer) GetNumPendingBlockRequests() int32 {
+	return peer.numPendingBlockRequests
+}
+
+func (peer *bpPeer) GetBlockAtHeight(height int64) (*types.Block, error) {
+	block, ok := peer.blocks[height]
+	if !ok {
+		return nil, errMissingRequest
+	}
+	if block == nil {
+		return nil, errMissingBlock
+	}
+	return peer.blocks[height], nil
+}
+
+func (peer *bpPeer) SetBlockAtHeight(height int64, block *types.Block) {
+	peer.blocks[height] = block
+}
+
+func (peer *bpPeer) RemoveBlockAtHeight(height int64) {
+	if _, ok := peer.blocks[height]; !ok {
+		return
+	}
+	delete(peer.blocks, height)
+}
+
+func (peer *bpPeer) resetMonitor() {
+	peer.recvMonitor = flow.New(peer.parameters.peerSampleRate, peer.parameters.peerWindowSize)
+	initialValue := float64(peer.parameters.minRecvRate) * math.E
+	peer.recvMonitor.SetREMA(initialValue)
+}
+
+func (peer *bpPeer) resetTimeout() {
+	if peer.blockResponseTimer == nil {
+		peer.blockResponseTimer = time.AfterFunc(peer.parameters.peerTimeout, peer.onTimeout)
+	} else {
+		peer.blockResponseTimer.Reset(peer.parameters.peerTimeout)
+	}
+}
+
+func (peer *bpPeer) IncrPending() {
+	if peer.numPendingBlockRequests == 0 {
+		peer.resetMonitor()
+		peer.resetTimeout()
+	}
+	peer.numPendingBlockRequests++
+}
+
+func (peer *bpPeer) StopBlockResponseTimer() bool {
+	if peer.blockResponseTimer == nil {
+		return false
+	}
+	return peer.blockResponseTimer.Stop()
+}
+
+func (peer *bpPeer) DecrPending(recvSize int) {
+	if peer.GetNumPendingBlockRequests() == 0 {
+		panic("cannot decrement, peer does not have pending requests")
+	}
+
+	peer.numPendingBlockRequests--
+	if peer.GetNumPendingBlockRequests() == 0 {
+		_ = peer.StopBlockResponseTimer()
+	} else {
+		peer.recvMonitor.Update(recvSize)
+		peer.resetTimeout()
+	}
+}
+
+func (peer *bpPeer) onTimeout() {
+	peer.errFunc(errNoPeerResponse, peer.id)
+}
+
+func (peer *bpPeer) IsGood() error {
```

Comment:
weird to see a method with `Is` prefix return an error (when you expect it to return boolean). Maybe add a comment?

Reply:
changed to CheckRate()